### ISSUE TITLE
false errors when using expected type and/or default type

### ISSUE THREAD

**Issue by KotlinIsland** [astral-sh/ty#3993]
### Summary

```py
from typing import Never

class A[T]:
    t: T

    def __add__[O, E=Never](self, other: A[O]) -> A[T | O | E]:
        return A()


a: A[int] = A()
b: A[int | None]
b = a + a  # Object of type `A[int]` is not assignable to `A[int | None]`
b = a.__add__(a)  # Argument to bound method `A.__add__` is incorrect: Expected `A[None | int]`, found `A[int]`
```

[Playground](https://play.ty.dev/84294bb1-ac8d-45bb-9205-86072a2f086c)

## and when we remove the default (that shouldn't affect this case at all)

```py
class A[T]:
    t: T

    def __add__[O, E](self, other: A[O]) -> A[T | O | E]:
        return A()


a: A[int] = A()
b: A[int | None]
b = a + a  # no error, but type is `A[int | Unknown]`
b = a.__add__(a)  # Argument to bound method `A.__add__` is incorrect: Expected `A[None | int]`, found `A[int]`
```

[Playground](https://play.ty.dev/210b6cb2-766c-4840-8cb3-3397d689310f)

**Comment by carljm** [comment]
This is the combination of #3420 (type context is not applied to implicit calls to binary dunders) and #3559 (return type context should participate as a constraint in the overall call solving, before fallbacks). Will keep it open as a useful demonstration case of the interaction between those. Thanks!
